hugr/src/proptest.rs

```diff
@@ -0,0 +1,122 @@
+use ::proptest::collection::vec;
+use ::proptest::prelude::*;
+use ::proptest::strategy::Strategy;
+use lazy_static::lazy_static;
+use smol_str::SmolStr;
+
+#[derive(Clone, Copy, Debug)]
+pub struct TypeDepth(usize);
+
+impl TypeDepth {
+    pub fn descend(&self) -> Self {
+        if self.leaf() {
+            *self
+        } else {
+            Self(self.0 - 1)
+        }
+    }
+
+    pub fn leaf(&self) -> bool {
+        self.0 == 0
+    }
+}
+
+impl Default for TypeDepth {
+    fn default() -> Self {
+        Self(3)
+    }
+}
+
+impl From<usize> for TypeDepth {
+    fn from(s: usize) -> Self {
+        Self(s)
+    }
+}
+
+lazy_static! {
+    static ref ANY_IDENT_STRING: SBoxedStrategy<String> = {
+        use proptest::string::string_regex;
+        prop_oneof![
+            string_regex(r"[[:alpha:]]{1,3}").unwrap(),
+            string_regex(crate::hugr::ident::PATH_COMPONENT_NICE_REGEX_STR).unwrap(),
+            string_regex(crate::hugr::ident::PATH_COMPONENT_REGEX_STR).unwrap(),
+        ].sboxed()
+    };
+
+    static ref ANY_NONEMPTY_STRING: SBoxedStrategy<String> = {
+        use proptest::string::string_regex;
+        prop_oneof![
+            string_regex(r"[[:alpha:]]{1,3}").unwrap(),
+            string_regex(r"[[:alpha:]]+").unwrap(),
+            string_regex(r".+").unwrap(),
+        ].sboxed()
+    };
+
+    static ref ANY_STRING: SBoxedStrategy<String> = {
+        use proptest::string::string_regex;
+        prop_oneof![
+            string_regex(r"[[:alpha:]]{0,3}").unwrap(),
+            string_regex(r"[[:alpha:]]*").unwrap(),
+            string_regex(r".*").unwrap(),
+        ].sboxed()
+    };
+
+    static ref ANY_SERDE_YAML_VALUE_LEAF: SBoxedStrategy<serde_yaml::Value> = {
```

Comment:
whoa, hang on. Surely we should not be constructing YAML this way. Arbitrary Rust object, serialize, validate against schema, deserialize, check equality ...?